client/src/partials/services/services.js

```diff
@@ -1,183 +1,145 @@
-      profit_center_id  : $scope.choosen.profit_center_id
-    };
-
-    if (isValid(data.cost_center_id, data.profit_center_id)) {
-      connect.put('service', [connect.clean(data)], ['id'])
-      .then(function () {
-        $scope.choosen.cost_center = $scope.choosen.cost_center_id ? getCostcenterText($scope.choosen.cost_center_id) : $scope.choosen.cost_center;
-        $scope.choosen.profit_center = $scope.choosen.profit_center_id ? getProfitCenterText($scope.choosen.profit_center_id) : $scope.choosen.profit_center;
-        $scope.model.services.put(connect.clean($scope.choosen));
-        $scope.action = '';
-        $scope.choosen = {}; // reset
-      })
-      .catch(function (err) {
-        messenger.danger('Error:' + JSON.stringify(err));
-      });
-    }else{
-      alert('centre de cout ou centre de profit deja utilise');
-    }
+    // load Profit Center
+    FinancialService.readProfitCenter().then(function (data) {
+      vm.profitCenters = data;
+    }).catch(handler);
 
+    vm.state.reset();
   }
 
-  function isValid (cost_id, profit_id) {
-    var c = $scope.model.services.data.filter(function (item) {
-      return (item.cost_center_id === cost_id && $scope.choosen.id !== item.id) || (item.profit_center_id === profit_id && $scope.choosen.id !== item.id);
-    })[0];
-
-    return c ? false : true;
+  function cancel() {
+    vm.state.reset();
+    vm.view = 'default';
   }
 
-  function getCostcenterText(cost_id) {
-    //FIX ME : hack
-    return configuration.cost_centers.filter(function (item) {
-      return item.id === cost_id;
-    })[0].text;
+  function create() {
+    vm.view = 'create';
+    vm.service = {};
   }
 
-  function getProfitCenterText (profit_id) {
-    //FIX ME : hack
-    return configuration.profit_centers.filter(function (item) {
-      return item.id === profit_id;
-    })[0].text;
+  // switch to update mode
+  function update(data) {
+    vm.state.reset();
+    vm.service= data;
+    vm.view = 'update';
   }
 
-
-  function handleResultCost(value) {
-    $scope.choosen.charge = value.data.cost;
-    return $q.when();
+  // switch to view more information about 
+  function more(data) {
+    vm.state.reset();
+    vm.service= data;
+    vm.choosen.service = data.name;
+    var ccId = data.cost_center_id;
+    var pcId = data.profit_center_id;
+    
+    // load Cost Center value for a specific service 
+    FinancialService.getCostCenter(vm.projectId,ccId).
+    then(function (data) {
+      vm.choosen.charge = data.cost;
+    }).catch(handler);
+
+    // load Profit Center value for a specific service 
+    FinancialService.getProfitCenter(vm.projectId,pcId).
+    then(function (data) {
+      vm.choosen.profit = data.profit;
+    }).catch(handler);
+
+    vm.view = 'more';
   }
 
-  function handleResultProfit(value) {
-    $scope.choosen.profit = value.data.profit;
-    return $q.when();
+  // switch to delete warning mode
+  function del(service) {
```

Comment:
Here is a better way to write this function
```js
function del(service) {
  var bool = $window.confirm($translate.instant('PROJECT.CONFIRM'));
  
   // if the user clicked cancel, reset the view and return
   if (!bool) {
      vm.view = 'default';
      return;
   }

  // if we get there, the user wants to delete a service
  vm.view = 'delete_confirm';
  Services.delete(service.id)
  .then(function () {
     vm.view = 'delete_success';
     return refershServices();
  })
  .catch(function (error) {{
    vm.HTTPError = error;
    vm.view = 'delete_error';
  });
}
```
This way of writing the function reduces complexity because:
 1. It returns early if the user click "cancel" on the alert
 2. It reduces the number of indented blocks in the code.
 3. Puts the asynchronous process (`Services.delete()`) as the last part of the function, helping the reader separate synchronous processes and asynchronous processes in their head.